parsec/mca/device/cuda/device_cuda_module.c

```diff
@@ -1385,7 +1396,15 @@ parsec_gpu_data_stage_in( parsec_device_cuda_module_t* cuda_device,
             PARSEC_DEBUG_VERBOSE(10, parsec_gpu_output_stream,
                                  "GPU[%s]:\tData copy %p [ref_count %d] on CUDA device %d is the best candidate to do Device to Device copy, increasing its readers to %d",
                                  gpu_device->super.name, candidate, candidate->super.super.obj_reference_count, target->cuda_index, candidate->readers+1);
-            parsec_atomic_fetch_inc_int32( &candidate->readers );
+            int readers = parsec_atomic_fetch_inc_int32(&candidate->readers);
+            if( readers < 0 ) {
+                PARSEC_DEBUG_VERBOSE(10, parsec_gpu_output_stream,
+                                     "GPU[%s]:\tCandidate %p [ref_count %d] on CUDA device %d is being repurposed by owner device. Looking for another candidate",
+                                     gpu_device->super.name, candidate, candidate->super.super.obj_reference_count, target->cuda_index);
+                /* We are trying to use a candidate that is repurposed by the owner device. Let's find another one */
+                parsec_atomic_fetch_add_int32(&candidate->readers, -1);
+                continue;
+            }
```

Comment:
After the increment we also need to check that the candidate has not been detached?

Reply:
That's the path if `readers < 0` (the case where we don't use the copy). But what about the alternative path (anything below this block)? We could encounter a copy with `readers == 1` that is detached from the data we're looking for, no?

Reply:
I don't think this scenario is possible anymore. Either the copy is not repurposed in which case the readers is `> 0`, or is currently repurposed by the owner in which case the atomicity guarantees that either readers is positive and the current thread can use it, is negative and the owner will get it from there.

However, as my comment in another of the comments states, if the current thread is significantly delayed at considering the copy, the owner could have completed the transfer and the copy is now associated with a different data_t. Without a lock I don't know how to avoid that.

Reply:
I think it is sufficient to check for the data after the increment (properly sequenced with a read memory barrier). There can be three cases:

1) `readers >= 0` and the right data: good, can use
2) `readers < 0`: bad, cannot use (another thread is in the process of detaching the copy)
3) `readers >= 0` and a different data: bad, cannot use (another thread has detached the copy and reset readers to 0)

Number 3 would be covered by checking for the data_t after the increment if the readers looked right, i.e., after this block. That is also why [the detach](https://github.com/ICLDisco/parsec/pull/575/files#diff-a42087f8b140e7fc55fd74480d49a4334d21ff2810f51ca57e0d1955534286e8R1072) needs to be properly sequenced with a write memory barrier (so that the detach cannot occur before the atomic operation on `readers`).